File: internal/domain/value_object/interval.go
package value_object

import (
	"beauty-server/internal/domain/errors"
	"time"
)

type Interval struct {
	Start time.Time
	End   time.Time
}

func NewInterval(start, end time.Time) (Interval, error) {
	if start.After(end) || start.Equal(end) {
		return Interval{}, errors.NewErrIntervalStartBeforeEnd(start, end)
	}
	return Interval{
		Start: start,
		End:   end,
	}, nil
}

func (i Interval) Overlaps(other Interval) bool {
	return i.Start.Before(other.End) && i.End.After(other.Start)
}

func (i Interval) Equal(other Interval) bool {
	return i.Start.Equal(other.End) && i.End.Equal(other.End)
}

File: Domain/ValueObjects/Interval.cs
using Domain.Errors;
using Domain.Primitives;
using Domain.Shared;
using System.Text.Json.Serialization;

namespace Domain.ValueObjects;

public class Interval : ValueObject
{
    public TimeSpan Start { get; }
    public TimeSpan End { get; }

    private Interval(TimeSpan start, TimeSpan end)
    {
        Start = start;
        End = end;
    }

    public static Result<Interval> Create(TimeSpan start, TimeSpan end)
    {
        if (start >= end)
        {
            return Result.Failure<Interval>(DomainErrors.TimeSlot.IntervalsOverlap);
        }

        return new Interval(start, end);
    }

    public bool Overlaps(Interval other)
    {
        return Start < other.End && End > other.Start;
    }

    public override IEnumerable<object> GetAtomicValues()
    {
        yield return Start;
        yield return End;
    }
}
